File: backend-quarkus/src/main/java/app/nanotes/backend/mail/Mailer.java
package app.nanotes.backend.mail;

import io.quarkus.mailer.Mail;
import jakarta.enterprise.context.ApplicationScoped;
import org.jboss.logging.Logger;

/**
 * Sends the share/mention notification emails required by the product
 * spec, via Quarkus's {@code quarkus-mailer} extension (configured from
 * {@code SMTP_*} env vars in application.properties).
 */
@ApplicationScoped
public class Mailer {

    private static final Logger LOG = Logger.getLogger(Mailer.class);

    private final io.quarkus.mailer.Mailer mailer;

    public Mailer(io.quarkus.mailer.Mailer mailer) {
        this.mailer = mailer;
    }

    /**
     * Logs failures instead of throwing: a notification email failing to
     * send must never fail the API request that triggered it (e.g. saving a
     * note the user just wrote).
     */
    private void sendBestEffort(String to, String subject, String body) {
        try {
            mailer.send(Mail.withText(to, subject, body));
        } catch (RuntimeException e) {
            LOG.errorf(e, "failed to send notification email to %s", to);
        }
    }

    public void notifyNoteShared(String to, String actorName, String noteTitle, String noteUrl, boolean editable) {
        String access = editable ? "editable" : "read-only";
        String subject = "%s shared a note with you: %s".formatted(actorName, noteTitle);
        String body = "%s shared the note \"%s\" with you (%s access).\n\nOpen it here: %s\n"
                .formatted(actorName, noteTitle, access, noteUrl);
        sendBestEffort(to, subject, body);
    }

    public void notifyMentioned(String to, String actorName, String noteTitle, String noteUrl) {
        String subject = "%s mentioned you in a note: %s".formatted(actorName, noteTitle);
        String body = "%s mentioned you in the note \"%s\".\n\nOpen it here: %s\n".formatted(actorName, noteTitle, noteUrl);
        sendBestEffort(to, subject, body);
    }
}

File: backend-quarkus/src/main/kotlin/app/nanotes/backend/mail/Mailer.kt
package app.nanotes.backend.mail

import io.quarkus.mailer.Mail
import jakarta.enterprise.context.ApplicationScoped
import org.jboss.logging.Logger

/**
 * Sends the share/mention notification emails required by the product
 * spec, via Quarkus's `quarkus-mailer` extension (configured from
 * `SMTP_*` env vars in application.properties).
 */
@ApplicationScoped
class Mailer(private val delegate: io.quarkus.mailer.Mailer) {

    companion object {
        private val LOG: Logger = Logger.getLogger(Mailer::class.java)
    }

    /**
     * Logs failures instead of throwing: a notification email failing to
     * send must never fail the API request that triggered it (e.g. saving a
     * note the user just wrote).
     */
    private fun sendBestEffort(to: String, subject: String, body: String) {
        try {
            delegate.send(Mail.withText(to, subject, body))
        } catch (e: RuntimeException) {
            LOG.errorf(e, "failed to send notification email to %s", to)
        }
    }

    fun notifyNoteShared(to: String, actorName: String, noteTitle: String, noteUrl: String, editable: Boolean) {
        val access = if (editable) "editable" else "read-only"
        val subject = "$actorName shared a note with you: $noteTitle"
        val body = "$actorName shared the note \"$noteTitle\" with you ($access access).\n\nOpen it here: $noteUrl\n"
        sendBestEffort(to, subject, body)
    }

    fun notifyMentioned(to: String, actorName: String, noteTitle: String, noteUrl: String) {
        val subject = "$actorName mentioned you in a note: $noteTitle"
        val body = "$actorName mentioned you in the note \"$noteTitle\".\n\nOpen it here: $noteUrl\n"
        sendBestEffort(to, subject, body)
    }
}
